src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs

```diff
@@ -291,16 +336,23 @@ public static RoutePattern Pattern(string rawText, params RoutePatternPathSegmen
                 {
                     if (segment.Parts[j] is RoutePatternParameterPart parameter)
                     {
+                        if (parameters == null)
+                        {
+                            parameters = new List<RoutePatternParameterPart>();
+                        }
+
                         parameters.Add(parameter);
                     }
                 }
             }
 
             return new RoutePattern(
                 rawText,
-                updatedDefaults,
-                updatedParameterPolicies.ToDictionary(kvp => kvp.Key, kvp => (IReadOnlyList<RoutePatternParameterPolicyReference>)kvp.Value.ToArray()),
-                parameters,
+                updatedDefaults ?? ReadOnlyDictionary<string, object>.Instance,
+                updatedParameterPolicies != null
+                    ? updatedParameterPolicies.ToDictionary(kvp => kvp.Key, kvp => (IReadOnlyList<RoutePatternParameterPolicyReference>)kvp.Value.ToArray())
+                    : ReadOnlyDictionary<string, IReadOnlyList<RoutePatternParameterPolicyReference>>.Instance,
```

Comment:
I don't think there's an issue with just caching a single empty (vanilla) dictionary and using it over and over. If you're worried about something introducing a bug, then use `ReadOnlyDictionary<,>` https://msdn.microsoft.com/en-us/ee572987

Reply:
Blah, somehow I missed that there was already an implementation in corefx. I thought it was ImmutableCollections only